### ISSUE TITLE
[HOLD for payment 2024-06-05] [$250] CRITICAL: Tapping on Push Notification fails to retrieve Onyx updates 

### ISSUE THREAD

**Issue by muttmuure** [Expensify/App#40227]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.4.62-3
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @muttmuure 
**Slack conversation:**

## Action Performed:
Tap a push notification

## Expected Result:
100% of the time, the message that generated that push should be rendered in the chat when you tap it

## Actual Result:
Since we removed `GetMissingOnyxMessages` from push and reconnect app, messages fail to be inserted into the client. 

## Workaround:
Opening chats and closing them retrieves the updates successfully, further suggesting that the command queue is to blame

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [X] Android: Native
- [X] Android: mWeb Chrome
- [X] iOS: Native
- [X] iOS: mWeb Safari
- [X] MacOS: Chrome / Safari
- [X] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fa6dcd0bbb47aa32</li>
        <li>Upwork Job ID: 1790746770223431680</li>
        <li>Last Price Increase: 2024-05-15</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @lschurr</details>

**Comment by muttmuure** [comment]
The proposed solution to this from @quinthar is:

1. Get a list of all updates
2. Sort them by updateID
3. Apply as many as you can in order
4. If there are any gaps, call GetMissingOnyxMessages once

**Comment by muttmuure** [comment]
https://github.com/Expensify/App/issues/38748

**Comment by lschurr** [comment]
On hold.

**Comment by lschurr** [comment]
This is on hold.

**Comment by quinthar** [comment]
#38748 is deployed; taking this off hold!

**Comment by chrispader** [comment]
I'll open up a PR to add these extra functions to the deferred updates Queue/`OnyxUpdateManager`.

**Comment by chrispader** [comment]
Thinking about the design of these additional encapsulated functions. When fetching the push notification updates with `Airship.push.getActiveNotifications()`, we receive the updates from all push notifications right?

I think it makes sense to automatically trigger the `validateAndApplyDeferredUpdates` process once deferred updates got queued. Otherwise we might incorporate some bad behaviors when deferred updates get queued but are not processed immediately.

Therefore, using `OnyxUpdateManager.enqueueDeferredUpdate()` would automatically trigger `processDeferredUpdates()`. I'll add an additional flag to this function, with which you can deactivate this automatic processing.

**Comment by arosiclair** [comment]
> Thinking about the design of these additional encapsulated functions. When fetching the push notification updates with `Airship.push.getActiveNotifications()`, we receive the updates from all push notifications right?
> 
> I think it makes sense to automatically trigger the `validateAndApplyDeferredUpdates` process once deferred updates got queued. Otherwise we might incorporate some bad behaviors when deferred updates get queued but are not processed immediately.
> 
> Therefore, using `OnyxUpdateManager.enqueueDeferredUpdate()` would automatically trigger `processDeferredUpdates()`. I'll add an additional flag to this function, with which you can deactivate this automatic processing.

Yup that all sounds right to me. I'll review the changes

**Comment by arosiclair** [comment]
Preliminary [deferred updates improvements](https://github.com/Expensify/App/pull/42044) are ready to be merged once the freeze is over. 

Next up is a PR for using it with push notifications. @chrispader would you be interested in working on that as well? You'll only be able to test Android push notifications but I can test iOS for you.

**Comment by arosiclair** [comment]
Not overdue

**Comment by lschurr** [comment]
Not overdue - not sure why the label is sticking around.

**Comment by chrispader** [comment]
> Next up is a PR for using it with push notifications. @chrispader would you be interested in working on that as well? You'll only be able to test Android push notifications but I can test iOS for you.

@arosiclair i can for sure work on it, but probably earliest next week, because of App.js Conf this week

**Comment by chrispader** [comment]
@arosiclair created a PR for this here: https://github.com/Expensify/App/pull/42510

I'm not sure if we actually have to use `DeferredOnyxUpdates` queue directly. Using `applyOnyxUpdatesReliably` synchronously should also achieve the same thing, while hiding internal code.

@arosiclair could you give the PR a look and let me know, if you think, that we actually need to use `DeferredOnyxUpdates` queue?

**Comment by chrispader** [comment]
i think @MelvinBot got this wrong. The solution has only been applied partially. Still waiting on https://github.com/Expensify/App/pull/42510 to be merged

**Comment by arosiclair** [comment]
Yeah this is not done but @allroundexperts should get a payment for his C+ review on [that first PR](https://github.com/Expensify/App/pull/42044). @lschurr

**Comment by arosiclair** [comment]
Alright upon further investigation and testing, I've found that we should _not_ add functionality to apply onyx updates when selecting a push notification. We have a `PushReceived` handler that applies the onyx update and this should work even when the app killed (the handler runs in the background OR the next time the app is re-opened). I confirmed that with Airship [here](https://expensify.slack.com/archives/C01RJ1RCK8D/p1717446337119679). So adding extra functionality on select would be redundant.

However, I did notice two other issues 

- Push notification Onyx updates are not applied correctly when there are no gaps detected. This is because the update is applied with a placeholder `eventType` [here](https://github.com/Expensify/App/blob/56424115c791d15bc44273c32a8c38e96f02165c/src/libs/Notification/PushNotification/subscribePushNotification/index.ts#L59) which results in a no-op since it's unrecognized [here](https://github.com/Expensify/App/blob/0140912da9f85a1eb17e7942a517c048c64b3c0f/src/libs/PusherUtils.ts#L19-L21). I'll fix this with a separate code flow for applying Airship onyx updates rather than re-using the pusher flow.
- Background refresh doesn't seem to be working on iOS anymore. The app doesn't wake in the background and only processes push notification events once you re-open the app.

I'm going to create two follow up issues to address those and leave this open since we still need to handle payment for the last C+ review

**Comment by arosiclair** [comment]
Follow up issues:
- https://github.com/Expensify/App/issues/43047
- https://github.com/Expensify/App/issues/43048

**Comment by lschurr** [comment]
The payment summary is correct. No payments will happen in Upwork. @allroundexperts you can request via NewDot :)

**Comment by lschurr** [comment]
Oh wait, should this one still be open @arosiclair? https://github.com/Expensify/App/issues/40227#issuecomment-2137469050

**Comment by arosiclair** [comment]
As long as [this payment](https://github.com/Expensify/App/issues/40227#issuecomment-2136925327) was handled then we should be all set

**Comment by lschurr** [comment]
Yep, that's handled in the payment summary here: https://github.com/Expensify/App/issues/40227#issuecomment-2150648610

Going to close :)

**Comment by JmillsExpensify** [comment]
$250 approved for @allroundexperts
